[debops.netbox] Create new Python 3 virtualenv

tumd opened the pull request:
Netbox v2.5 and newer no longer supports Python 2.
This creates a new vituealenv directory based on Python 3 when
`netbox__git_version` is v2.5+ or master/develop.
This *should* make a clean upgrade of existing Netbox installations.

I'm not sure this is the best way to handle this kind of situation. I'm not that 'at home' with python virtualenvs, and couldn't find a way to upgrade the existing virtualenv to python 3.
This also most likely needs a bit more testing.

tumd committed:
[debops.netbox] Create new Python 3 virtualenv

Netbox v2.5 and newer no longer supports Python 2.
This creates a new vituealenv directory based on Python 3 when
netbox__git_version is v2.5+ or master/develop.
This *should* make a clean upgrade of existing Netbox installations.

drybjed commented:
@tumd thanks, it looks OK to me, although I'll test it a few times with existing and new NetBox installation before merging, to be sure that it works as expected.

tumd committed:
[debops.netbox] Added some missing packages

tumd commented:
Don't merge.
This doesn't work in the current state.
Will have another look at it in the upcoming days.

tumd committed:
[debops.netbox] Remove the old python2 virtualenv

If a virtualenv based on python3 is needed, remove the old one
entirely.
This will include checking out the nextbox-repo anew.

tumd committed:
Merge branch 'master' into netbox-python3

tumd commented:
With the last commit I changed the behavior slightly. The old virtualenv will now be removed and recreated if it's outdated (i.e. python2 based).
I'm not entirely happy with how `netbox__virtualenv_version` is used, but I wasn't sure how else the version-change of packages and others should be handled.

Please try it out and let me know what you think.

drybjed commented on ansible/roles/debops.netbox/defaults/main.yml:
```diff
@@ -40,15 +40,15 @@ netbox__domain: '{{ ansible_local.core.domain
-  - 'python-dev'
-  - 'python-pip'
+  - 'python{{ netbox__virtualenv_version }}-dev'
+  - 'python{{ netbox__virtualenv_version }}-pip'
```
The `netbox.yml` playbook already contains the `debops.python` role which is required by the `debops.gunicorn` role. You could move the `python*` APT packages to the `netbox__python__dependent_packages2` and `netbox__python__dependent_packages3`, so that the `debops.python` role itself installs the desired version (see how it's done in `debops.gunicorn` role for an example). You would probably only need to install `python-dev`/`python3-dev` packages since `pip` and `virtualenv` is already included in `debops.python` role by default.

tumd replied:
This should now be implemented.

drybjed reviewed (commented):
Changes look promising, but I haven't tested them yet, will do during the weekend.

I gave you a possible better solution for installing APT `python` packages via `debops.python` role. You could try implementing it if you want, or we can do that later after the PR is merged.

tumd committed:
[debops.netbox] Moved python-deps to debops.python

Also fixed the restart handler for gunicorn@netbox

tumd commented:
The changes has been tested both from a clean install and by upgrading a existing netbox installed using the current debops/master (with sample-data only). Seems to be working.

The existing tasks handling db-migration and such seem to be working too.
However, netbox provides their own [upgrade.sh](https://github.com/digitalocean/netbox/blob/develop/upgrade.sh) to handle upgrades, which pretty much do the same thing as many of the tasks in this role. After a quick look in the git-history, it looks like the role was created before the upgrade.sh-script existed upstream.
Would it be a good idea to try to utilize this script instead of the tasks?

drybjed commented:
In the past, the `upgrade.sh` script used `sudo` commands internally, therefore it was unsuitable for installation managed by DebOps in a virtualenv. This has been fixed about a month ago when use of `sudo` inside of the script was dropped. So I guess now the `debops.netbox` role could be switched over to it after checking that everything works as expected. But, I would wait with this change a bit to let people update to newer NetBox version, and perhaps do it after new DebOps version is released.

tumd committed:
[debops.netbox] Real fix for gunicorn restart

drybjed reviewed (approved):
I tested these changes on an existing NetBox installation and upgrade went smoothly, everything seems to be in order. Nicely done! :+1:

tumd commented:
Sweet!
This probably also close #580
